modules/xphoto/src/grayworld_white_balance.cpp

```diff
@@ -0,0 +1,213 @@
+        v_uint16x8 v_s1, v_s2;
+        v_uint32x4 v_iB1, v_iB2, v_iB3, v_iB4,
+                   v_iG1, v_iG2, v_iG3, v_iG4,
+                   v_iR1, v_iR2, v_iR3, v_iR4,
+                   v_255 = v_setall_u32(255),
+                   v_thresh = v_setall_u32(thresh255),
+                   v_min1, v_min2, v_min3, v_min4,
+                   v_max1, v_max2, v_max3, v_max4,
+                   v_m1, v_m2, v_m3, v_m4,
+                   v_SB = v_setzero_u32(),
+                   v_SG = v_setzero_u32(),
+                   v_SR = v_setzero_u32();
+
+        for ( ; i < N3 - 47; i += 48 )
+        {
+            // NOTE: This block assumes BGR channels in naming variables
+
+            // Load 3x uint8x16 and deinterleave into vectors of each channel
+            v_load_deinterleave(&src_data[i], v_inB, v_inG, v_inR);
+
+            // Split into four int vectors per channel
+            v_expand(v_inB, v_s1, v_s2);
+            v_expand(v_s1, v_iB1, v_iB2);
+            v_expand(v_s2, v_iB3, v_iB4);
+
+            v_expand(v_inG, v_s1, v_s2);
+            v_expand(v_s1, v_iG1, v_iG2);
+            v_expand(v_s2, v_iG3, v_iG4);
+
+            v_expand(v_inR, v_s1, v_s2);
+            v_expand(v_s1, v_iR1, v_iR2);
+            v_expand(v_s2, v_iR3, v_iR4);
+
+            // Get mins and maxs
+            v_min1 = v_min(v_iB1, v_min(v_iG1, v_iR1));
+            v_min2 = v_min(v_iB2, v_min(v_iG2, v_iR2));
+            v_min3 = v_min(v_iB3, v_min(v_iG3, v_iR3));
+            v_min4 = v_min(v_iB4, v_min(v_iG4, v_iR4));
+
+            v_max1 = v_max(v_iB1, v_max(v_iG1, v_iR1));
+            v_max2 = v_max(v_iB2, v_max(v_iG2, v_iR2));
+            v_max3 = v_max(v_iB3, v_max(v_iG3, v_iR3));
+            v_max4 = v_max(v_iB4, v_max(v_iG4, v_iR4));
+
+            // Calculate masks
+            v_m1 = ~((v_max1 - v_min1) * v_255 > v_thresh * v_max1);
+            v_m2 = ~((v_max2 - v_min2) * v_255 > v_thresh * v_max2);
+            v_m3 = ~((v_max3 - v_min3) * v_255 > v_thresh * v_max3);
+            v_m4 = ~((v_max4 - v_min4) * v_255 > v_thresh * v_max4);
+
+            // Apply mask
+            v_SB += (v_iB1 & v_m1) + (v_iB2 & v_m2) + (v_iB3 & v_m3) + (v_iB4 & v_m4);
+            v_SG += (v_iG1 & v_m1) + (v_iG2 & v_m2) + (v_iG3 & v_m3) + (v_iG4 & v_m4);
+            v_SR += (v_iR1 & v_m1) + (v_iR2 & v_m2) + (v_iR3 & v_m3) + (v_iR4 & v_m4);
+        }
+
+        // Perform final reduction
+        sum1 = v_reduce_sum(v_SB);
+        sum2 = v_reduce_sum(v_SG);
+        sum3 = v_reduce_sum(v_SR);
+#endif
+        unsigned int minRGB, maxRGB;
+        for ( ; i < N3; i += 3 )
+        {
+            minRGB = min(src_data[i], min(src_data[i + 1], src_data[i + 2]));
+            maxRGB = max(src_data[i], max(src_data[i + 1], src_data[i + 2]));
+            if ( (maxRGB - minRGB) * 255 > thresh255 * maxRGB ) continue;
+            sum1 += src_data[i];
+            sum2 += src_data[i + 1];
+            sum3 += src_data[i + 2];
+        }
```

Comment:
As `dinv#` are divided by `max(dinv1, dinv2, dinv3)`, the scaling factor does not actually change the result if I understand correctly.

I was unsure if it would be worth adding a `v_reduce_sum` call into vectorised version to keep the concept of averaging correct.